datafusion/expr/src/signature.rs

```diff
@@ -132,7 +138,136 @@ pub enum ArrayFunctionSignature {
     /// The first argument should be non-list or list, and the second argument should be List/LargeList.
     /// The first argument's list dimension should be one dimension less than the second argument's list dimension.
     ElementAndArray,
+    /// Specialized Signature for ArrayEqual and similar functions
     ArrayAndIndex,
+    /// Specialized Signature for ArrayEmpty and similar functions
+    Array,
+}
+
+impl ArrayFunctionSignature {
+    /// Arguments to ArrayFunctionSignature
+    /// `current_types` - The data types of the arguments
+    /// `allow_null_coercion` - Whether null type coercion is allowed
+    /// Returns the valid types for the function signature
+    pub fn get_type_signature(
+        &self,
+        current_types: &[DataType],
+        allow_null_coercion: bool,
```

Comment:
since we are passing allow_null_coercion here, maybe it would make sense to have allow_null_coercion to be part of each array enum? Perhaps something like

```rust
impl ArrayFunctionSignature {
    ArrayAndElement(allow_null_coercion),
    ...
    Array(allow_null_coercion)
}
```

But again I am not sure about the need for allow_null_coercion 🤔

Reply:
`allow_null_coercion` determines if the signature can allow a `null` type. For example, `empty(NULL)` accepts 		`NULL` while `array_dims(Null)` will return error